services/telemetry/datasource_pmmdb_select.go

```diff
@@ -0,0 +1,99 @@
+// pmm-managed
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package telemetry provides telemetry functionality.
+package telemetry
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"net/url"
+
+	pmmv1 "github.com/percona-platform/saas/gen/telemetry/events/pmm"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type dsPmmDBSelect struct {
+	l      *logrus.Entry
+	config DSConfigPMMDB
+	db     *sql.DB
+}
+
+// check interfaces
+var (
+	_ DataSource = (*dsPmmDBSelect)(nil)
+)
+
+// Enabled flag that determines if data source is enabled.
+func (d *dsPmmDBSelect) Enabled() bool {
+	return d.config.Enabled
+}
+
+// NewDsPmmDBSelect make new PMM DB Select data source.
+func NewDsPmmDBSelect(config DSConfigPMMDB, l *logrus.Entry) (DataSource, error) { //nolint:ireturn
+	db, err := openPMMDBConnection(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &dsPmmDBSelect{
+		l:      l,
+		config: config,
+		db:     db,
+	}, nil
+}
+
+func openPMMDBConnection(config DSConfigPMMDB) (*sql.DB, error) {
```

Comment:
looks like this function duplicates `models.OpenDB`

Reply:
as far as I see difference in two lines of code which can be overriden by passing new params or by changing them once `db` is returned.